Define IBC communication (#49)
osmosis-labs/mesh-security: docs/ibc/Staking.md

by the provider is asigning virtual stake to particular validator and removing it.

It does so via `Stake` and `Unstake` messages, specifying the valoper address it wishes
to delegate to and the amount of local (provider-side, held in the vault) tokens that
it wishes to virtually stake.

The converter must track the total sum of these messages for each validator on the consumer
side, convert them to proper values of native tokens and "virtual stake" them.

## Channel Ordering

Note the entire protocol is designed around syncing an initial state and sending a stream
of diffs, such that `State = Initial + Sum(Diffs)`. This applies both the validator set
as well as the total amounts staked. 

If we reorder the diffs, it is possible to get a different result, so we need to be careful
about relying on Unordered channels. Imagine Stake 100 and Unstake 50. If Unstake goes first,
it would return an error, yet the Stake would apply properly, leaving a total of 100 not 50.
Furthermore, if a packet is dropped, and the diff is not applied, the two sides will
get out of sync, where eg the Provider believes there is 500k staked to a given validator,
while the Consumer believes there is 400k. 

> At the same time, Ordered channels are fragile, in that a single timed-out or undeliverable packet

Author: Yup.

CRDTs are not possible as they say every ordering should give the same final state. Which makes enforcing invariants like "never go below 0" impossible unless it is a monotonically increasing function.

If +10 -5  is valid, then a CRDT would require -5 +10 to also be valid and have the same meaning.
Yet this would lead to a negative stake error. 

I was really into CRDTs for better data syncing (between devices with spotty coverage) in my last job before blockchain. But it doesn't really work in blockchain, as it really asserts ordering and "always >=  0" invariants to avoid "double spend attack".

> At the same time, Ordered channels are fragile, in that a single timed-out or undeliverable packet

Author: We can be inspired by vector clocks, but again, there is no merge function here. There is one state.

We can imagine all possible paths / orderings and then ensure that they all lead to consistent states, but possibly different ones.

> At the same time, Ordered channels are fragile, in that a single timed-out or undeliverable packet

Author: I spent some time last night reflecting more and realize the *validator set* is actually a perfect use of CRDT. 

We just have to treat `Add(A), Remove(A)` the same as `Remove(A), Add(A)`, meaning not error on removing a non-existant validator, but store this drop message for later.

a staking action disappear with no notification as to why will lead to serious confusion
and cries of "bugs", when it was just an invisible, delayed error.

An alternate approach here is to have some sort of "re-sync" design to re-synchronize the
state on the two sides, but I argue this is too fragile in an asynchrnous environment.
Much care should be taken to ensure that all applications errors are handles well.

### Error handling

When staking, create a `Stake` packet, but do not update the user's / validator's state
in the `external-staking` contract. If an error ack comes back, then the
`external-staking` contract should immediately call back to the vault to release
the lein for the amount of that packet. If a success ack comes back, it should
actually apply the staking changes locally.

Likewise, when unstaking, we first create an `Unstake` packet. However, here 
we want to avoid the situation of `Unstaking` 100% of the tokens multiple times,
as this would apply invalid state changes to the consumer. We need to enforce this 
limit on the provider side (which is handled by the vault in the staking case).

> For `Unstake`, we should update a local "unstaking" value on the `(user, validator)`

Author: Not needed there. We need it for unstake to avoid going below 0, as we track how much can be unstaked in this contract.

For staking, the assurance that they haven't staked more than they have collateral is enforced in the vault. This is out "pending_staked" in a way. The change *is* written to vault before we try to send IBC message, which is why it is important to revert that by releasing the lien if the packet fails.

I don't see the need for tracking this a second time in the external-stake

> For `Unstake`, we should update a local "unstaking" value on the `(user, validator)`

Reviewer: Fair enough, when this is tracked in the vault. The user should be prohibited to stake the same tokens twice while it is in-flight.

> For `Unstake`, we should update a local "unstaking" value on the `(user, validator)`

Author: Yes, very good point. I will elaborate on how we enforce this

staking info, but not create a claim nor apply a diff to the validator. 

> We ensure that this "unstaking" amount can never be larger than the properly staked

Author: Can you expand on this point?

> We ensure that this "unstaking" amount can never be larger than the properly staked

Reviewer: Consumer and. provider side can become out of sync on the staked amount when slashing happens.
Slashing is originated on the consumer side and reduces the effectively staked amount.  It will take some time to propagate the event to the provider side. During this period both sides are out of sync and the real stake amount is < provider side stake amount.

> We ensure that this "unstaking" amount can never be larger than the properly staked

Author: Oh, that is a very good point. We need a much longer section on slashing. But I will include a space to dig into this point more.

(and ack'ed) value. On an error ack, we simply reduce "unstaking" by that amount.
On a success ack, we commit these changes, reducing not only "unstaking", but
also applying the deduction to actual stake for the user as well as the validator,
updating the rewards claim table, and creating the claim, so the user can get their
tokens / lein back after the unbonding period.

### Re-syncing

We could invent some "re-sync" mechanism, but would have to be careful mixing this with
in-flight messages. For example, the provider sends a `Stake` message to the consumer,
which returns an error for whatever reason. Before the ack has arrived, the provider
sends a re-sync message with the entire content of it's local state.

Upon receiving that "re-sync" message, the consumer updates all tables and triggers the
appropriate virtual staking commands. However, the error ack for the `Stake` packet
now lands on the provider and it "undoes" the Stake action. This will be applied

> on top of the snapshot that it sent to the consumer, thus modifying it and the two sides

Reviewer: I got a bit lost in technical details. I think if we can switch from push (consumer) to poll (provider side) as start of the process, things should just work on an ordered channel. The provider would ask for a state sync on validator(s) and then receive the result. Any other operation started by the provider side should be completed before the sync arrives, except for slashing.

# Cross-Chain Staking Protocol

The staking protocol has two basic operations:

* Stake X tokens on Validator V - `S(V, X)`
* Unstake X tokens from Validator V - `U(V, X)`

We want to ensure that at any point in time, when all in-flight messages would be resolved, both the
provider and the consumer chain have the same view of the staking state. (**TODO** Consider the
effects of slashing, not included in this document currently). 

We also want to guarantee that the provider chain always maintains sufficient staked tokens
in the vault to cover all virtual staking actions currently outstanding on the provider chain.

As [mentioned before](./ControlChannel.md#channel-ordering), we wish to use an unordered channel,

> and therefore must bring a degree of understading of [Serializability](./Serializability.md)

Reviewer 2: understading -> understanding

to this protocol.

## Delegation Syncing

The consumer side wants to maintain the same counter per validator as the provider side.
It manages a delegation count per validator - `D[V]`, with the following rules:

* Uninitialized is equivalent to `D[V] = 0`
* `S(V, X)` => `D[V] += X`, return success ack
* `U(V, X)` => 
  * if `D[V] < X`, return error ack
  * else `D[V] -= X`, return success ack

> This is pretty straightforward counter with a lower bound of 0, along with an increment and decrement

Reviewer 2: ```suggestion
This is a pretty straightforward counter with a lower bound of 0, along with an increment and decrement
```

counters. 

The only requirements to ensure this stays in sync is that the provider can successfully commit the same

> changes upon a success ack, and is able to revert then (unstake) on an error ack. 

Reviewer 2: then -> them

**TODO** Do we need text / code / explanation here?

## Provider Side Design

Beyond this delegation list, the provider must maintain much more information.
The delegation to a given validator must be mapped to a user on the provider chain.
This user must also have a sufficient lien on the vault to cover the delegation.
And the vault must have sufficient funds to cover the lien.

This means we must maintain invariants over at least two contracts - the `external-staking` contract

> and the `vault`. It includes possible conflicts with non-IBC transactions, like a user to withdrawing 

Reviewer 2: Remove 'to'.

  * Ensure there is sufficient collateral in sum(potential slashes)
  * Increase the lien for that given user on the `external-staking` contract
* Add a delegation in `external-staking` contract
  * Increase stake count for (user, validator)
  * Increase total stake on the validator
  * Increase the user's shares in the reward distribution
* Send an IBC packet to inform the Consumer
  * Guarantee we can commit all above on success
  * Guarantee we can rollback all above on error

An unstaking operation would have the following steps and checks:

* Remove a delegation in `external-staking` contract
  * Ensure stake count (user, validator) is set and greater than desired unstake amount
    * Ensure total stake on the validator is set and greater than desired unstake amount (should always be true if above is true)
  * Decrease stake count for (user, validator)
  * Decrease total stake on the validator
  * Decrease the user's shares in the reward distribution
  * Add an entry to the (user, validator) "pending unbonding" to withdraw said tokens after the unbonding period.
* Send an IBC packet to inform the Consumer
  * Guarantee we can commit all above on success

>   * Guarantee we can rollback all above on error

Reviewer 2: Looks good, but we must be really sure we're not missing some scenario / combination of operations here.

Perhaps a simpler initial approach could be, to allow for only one in-flight operation at any given time, and reject new operations until the pending one is completed, either successfully or with error.

Author: That is the locking approach.

We can't just reject other IBC packets, but any other transaction that touches the same keys.
That means, we cannot pay out rewards to that validator or make/remove another lien from that user to anyone.

That is correct, but not efficient. Please read discussion in Serializability and we take this up.

The general process (assuming a vault is already established on the provider) is:

1. Instantiate price feed, converter, virtual staking contracts on the consumer chain
2. Instantiate external staking contract on the provider chain (referencing IBC port of the converter)
3. Create IBC channel from provider to consumer
4. Apply to consumer governance to provide a virtual staking max cap to the associated virtual staking contract, so that this connection may have voting power.

### Version Negotiation

The channel version uses a JSON-encoded struct with the following fields:

```json
{
    "protocol": "mesh-security",
    "version": "1.0.0",
}
```

It is important that you **do not** use `#[cw_serde]` on the Rust struct as we explicitly
want to **allow unknown fields**. This allows us to add new fields in the future.
`#[cw_serde]` generates `#[serde(deny_unknown_fields)]` which would break this.

Both sides must error if the protocol is not `mesh-security`. 

The version is used to allow for future upgrades. The provider should always send the
highest protocol version it supports to start the handshake. The consumer should

> error if the major version is higher than it's known versions (for now, anything besides `1`).

Reviewer 2: it's -> its

A contract panic will abort the tx containing the IbcPacketReceiveMsg as of wasmd 0.40
(MSV for Mesh Security)

## Implementation Notes

In order to maintain state synchronized on both sides, we must ensure that the
sending side properly handle acks, both success and failure. 

If there is a failure sending a validator set update, it should be retried later.
A safe way to do so would be to store these messages in some "queue" and trigger sending 
such `AddValidator` and `RemoveValidator` packets on the next incoming Stake/Unstake command
(when we know the provider side is working). They should be re-sent in the same order
they were originally sent.

If there is a failure sending a Stake / Unstake action, the safest response is to
undo the action locally and inform the user. We need solid UI here, as suddenly having
a staking action disappear with no notification as to why will lead to serious confusion
and cries of "bugs", when it was just an invisible, delayed error.

An alternate approach here is to have some sort of "re-sync" design to re-synchronize the
state on the two sides, but I argue this is too fragile in an asynchrnous environment.

> Much care should be taken to ensure that all applications errors are handles well.

Reviewer 2: ```suggestion
Much care must be taken to ensure that all application errors are handled well.
```

### Error handling

When staking, create a `Stake` packet, but do not update the user's / validator's state
in the `external-staking` contract. If an error ack comes back, then the
`external-staking` contract should immediately call back to the vault to release

> the lein for the amount of that packet. If a success ack comes back, it should

Reviewer 2: lein -> lien

- Unstake X tokens from Validator V - `U(V, X)`

We want to ensure that at any point in time, when all in-flight messages would be resolved, both the
provider and the consumer chain have the same view of the staking state. (**TODO** Consider the
effects of slashing, not included in this document currently).

We also want to guarantee that the provider chain always maintains sufficient staked tokens
in the vault to cover all virtual staking actions currently outstanding on the provider chain.

As [mentioned before](./ControlChannel.md#channel-ordering), we wish to use an unordered channel,
and therefore must bring a degree of understanding of [Serializability](./Serializability.md)
to this protocol.

## Delegation Syncing

The consumer side wants to maintain the same counter per validator as the provider side.
It manages a delegation count per validator - `D[V]`, with the following rules:

- Uninitialized is equivalent to `D[V] = 0`
- `S(V, X)` => `D[V] += X`, return success ack
- `U(V, X)` =>
  - if `D[V] < X`, return error ack
  - else `D[V] -= X`, return success ack

This is a pretty straightforward counter with a lower bound of 0, along with an increment and decrement

> counters.

Author: I am very hesitant to relax invariants ever, and start handling negative numbers when an epoch hits.

The provider should not even allow the `U(a, 50)` action to be initiated on that provider side, until it is sure the `D(a, 100)` is committed. What happens if the first one fails for some reason?

With full CRDTs, we assume nothing will ever be rejected, and all operations will be applied.

With this distributed transaction model, we need to ensure consistency. Minimally, no transaction can read data from an in-flight transaction - meaning the delegation doesn't even start on the provider side until it is fully commited (on ack not on send packet)

We can simplify this list by making use of our knowledge of how various states are derived from each other:

- `total-stake(validator)` is the sum of `stake(user, validator)` over all users.
- `reward-shares(validator, user)` is the same as `stake(user, validator)`.
- `total-shares(validator)` is the sum of `reward-shares(validator, user)` over all users.

By analyzing how these other values are derived, we can see that if `stake(user, validator)` is never negative, then
none of the other values can ever be negative either. This means we can simplify those four keys to simply lock `stake(user, validator)`,
which is sufficient to guarantee the invariants of the other staking values, without interfering with actions by other users.

We end up with two keys that need to write locks, both of which only affect a single user:

- `vault::lien(user, external-staking)`
- `external-staking::stake(user, validator)`

Note that a write-lock will prevent reading of the value from any other transaction. That means the same user offering a lien on another
validator, which iterates over all liens to find the sum of max slashing, will be blocked until the first transaction completes.

> We cannot actually block (or wait) transactions in this way, so we must return an error for the second transaction.

Reviewer: I think that you can avoid failing the second TX by maintaining state for in flight data in both contracts.
Like credit-card companies, you could place a lock in the vault for the lien amount, that is released and the change applied on success Ack or just dropped on error Ack. The liquid balance would be reduced by the locked amounts.
Similar in external-staking, where the changes are applied only on success Ack. Or reverted on error Ack. Depending at which stage they are taken into account in the contract

Reviewer: I may not have the full picture, yet but with my current understanding, the "withdraw collateral" should fail instead. The "stake" action was started before and the vault should reflect that state in the balance. With an optimistic approach, the lien should be taken on start of the "stake" process already. It can be returned on error Ack or timeout. 

I do understand the "write-lock" as "one operation only until finished". That constraint would be very hard and can be avoided IMHO. It should be possible to "withdraw collateral" and having a "stake" operation in flight as long as the balance in the vault is sufficient for both.

Author: Yes, the locking approach is the easiest one to prove safe.

And from there, I look to reduce constraints, while showing that this can never open up a conflict, assuming the previous version had no conflicts.

I think you are not looking at the latest version, which has a section **Idea: Approximating Locks** discussing just what you are suggesting (and asking for feedback that this is safe as a lock)

The staking protocol has two basic operations:

- Stake X tokens on Validator V - `S(V, X)`
- Unstake X tokens from Validator V - `U(V, X)`

We want to ensure that at any point in time, when all in-flight messages would be resolved, both the
provider and the consumer chain have the same view of the staking state. (**TODO** Consider the
effects of slashing, not included in this document currently).

We also want to guarantee that the provider chain always maintains sufficient staked tokens
in the vault to cover all virtual staking actions currently outstanding on the provider chain.

As [mentioned before](./ControlChannel.md#channel-ordering), we wish to use an unordered channel,
and therefore must bring a degree of understanding of [Serializability](./Serializability.md)
to this protocol.

## Delegation Syncing

The consumer side wants to maintain the same counter per validator as the provider side.
It manages a delegation count per validator - `D[V]`, with the following rules:

- Uninitialized is equivalent to `D[V] = 0`
- `S(V, X)` => `D[V] += X`, return success ack
- `U(V, X)` =>
  - if `D[V] < X`, return error ack
  - else `D[V] -= X`, return success ack

> This is a pretty straightforward counter with a lower bound of 0, along with an increment and decrement

Reviewer 2: Perhaps reference here the upper bound (max cap) as well?

This is another issue, but I think we would need a message type for max cap to reach the provider side. That way we can know how much it can be delegated to a given remote, before sending the delegation.
That also involves converting between provider and consumer coins. So, it will have to be updated once per day or so, based on the price oracle.

Or am I mistaken here?

Author: We store the total requested stake in the consumer side.

When the epoch comes, it is normalized down to max_cap if above, and then redelegated appropriately.

Eg. if max cap is 1000 and there was 500 on A and 500 B. Then I stake 1000 more on A. When the epoch tick comes, it doesn't end up 500-500 but rather 750 on A and 250 on B (staked * max cap / total staked)

Author: We let the provider do what it wants, it always has impact.
But after max_cap, there is diminishing returns per staker in terms of rewards, and they come from other cross stakers on the same channel

  - Guarantee we can rollback all above on error

An unstaking operation would have the following steps and checks:

- Remove a delegation in `external-staking` contract
  - Ensure stake count (user, validator) is set and greater than desired unstake amount
    - Ensure total stake on the validator is set and greater than desired unstake amount (should always be true if above is true)
  - Decrease stake count for (user, validator)
  - Decrease total stake on the validator
  - Decrease the user's shares in the reward distribution
  - Add an entry to the (user, validator) "pending unbonding" to withdraw said tokens after the unbonding period.
- Send an IBC packet to inform the Consumer
  - Guarantee we can commit all above on success
  - Guarantee we can rollback all above on error

Possible keys with conflicts are:

- In `vault` - `collateral(user)` and `lein(user, external-staking)`
- In `external-staking` - `stake(user, validator)`, `total-stake(validator)`, `reward-shares(user)`, `pending-unbonding(user, validator)`

### Identifying Potential Commutatibility

The general design should be to write all changes only on a successful ACK, but hold any locks needed to ensure those

> write will not fail in any condition. Using the apporach of [Value Ranges](./Serializability.md#value-ranges), let us analyze

Reviewer 2: write -> writes
apporach -> approach

what needs to be minimally enforced here.

For staking, we update:

- `vault::lien(user, external-staking)` => `Maybe(+X)`
- `external-staking::stake(user, validator)` => `Maybe(+X)`
- `external-staking::total-stake(validator)` => `Maybe(+X)`
- `external-staking::reward-shares(validator, user)` => `Maybe(+X)`

> The lower three have no upper limit and thus can never fail, so those operations are always commutative with others valid operations.

Author: Well, when lots of people ape in to stake somewhere, APRs drop.
This is true all over DeFi and already (to a lesser extend) in staking rewards (in the early days of chains).
Users should be aware (or have a little disclaimer on the app)

- `total-stake(validator)` is the sum of `stake(user, validator)` over all users.
- `reward-shares(validator, user)` is the same as `stake(user, validator)`.
- `total-shares(validator)` is the sum of `reward-shares(validator, user)` over all users.

By analyzing how these other values are derived, we can see that if `stake(user, validator)` is never negative, then
none of the other values can ever be negative either. This means we can simplify those four keys to simply lock `stake(user, validator)`,
which is sufficient to guarantee the invariants of the other staking values, without interfering with actions by other users.

We end up with two keys that need to write locks, both of which only affect a single user:

- `vault::lien(user, external-staking)`
- `external-staking::stake(user, validator)`

Note that a write-lock will prevent reading of the value from any other transaction. That means the same user offering a lien on another
validator, which iterates over all liens to find the sum of max slashing, will be blocked until the first transaction completes.
We cannot actually block (or wait) transactions in this way, so we must return an error for the second transaction.

> ### Idea: Approximating Locks

Reviewer 2: Can't review now, but will take a look later for sure.

> ### Idea: Approximating Locks

Author: @alpe Your comments above about optimisitic liens are the topic of this section.

Please read it and let me know how it can be improved.

Packet along with the ack and be able to properly commit/rollback the staking action.

### Consumer-Side Logic

This is quite simple, as the Consumer only needs to apply the staking action to its local state, and return an error
if this violates some invariant (goes below 0):

```rust
match op {
    Stake { validator, user, amount} => {
        ensure_eq!(amount.denom, cfg.staking_denom, IbcError::InvalidDenom);
        let delegated = DELEGATIONS.may_load(store, validator)?.unwrap_or_default();
        delegated += amount.amount;
        DELEGATIONS.save(store, validator, delegated)?;
    }
    Unstake { validator, user, amount} => {
        ensure_eq!(amount.denom, cfg.staking_denom, IbcError::InvalidDenom);
        /// They must have a previous delegation to unstake
        let delegated = DELEGATIONS.load(store, validator)?.checked_sub(amount.amount)?;
        DELEGATIONS.save(store, validator, delegated)?;
    }
}
```

### Provider-Side Staking

This is more complex logic over multiple contracts, so I will only give a high-level overview here rather the Rust psuedo-code:

**TODO**

> See [open question above](#idea-approximating-locks) for discussion of how to possibly avoid locking on staking.

Reviewer 2: ```suggestion
See [open question above](#idea-approximating-locks) for a discussion on how to possibly avoid locking on staking.
```
